ext/SummationByPartsOperatorsOptimForwardDiffExt.jl

```diff
@@ -0,0 +1,239 @@
+    S_cache = DiffCache(S)
+    A_cache = DiffCache(A)
+    SV_cache = DiffCache(SV)
+    PV_x_cache = DiffCache(PV_x)
+    daij_dsigmak = zeros(eltype(nodes), N, K, L)
+    daij_drhok = zeros(eltype(nodes), N, K, N)
+    p = (V, V_x, R, x_length, S_cache, A_cache, SV_cache, PV_x_cache, daij_dsigmak, daij_drhok)
+    # @views function optimization_function(x, p)
+    #     V, V_x, R, x_length, S_cache, A_cache, SV_cache, PV_x_cache = p
+    #     S = get_tmp(S_cache, x)
+    #     A = get_tmp(A_cache, x)
+    #     SV = get_tmp(SV_cache, x)
+    #     PV_x = get_tmp(PV_x_cache, x)
+    #     (N, _) = size(R)
+    #     L = Integer(N*(N - 1)/2)
+    #     sigma = x[1:L]
+    #     rho = x[(L + 1):end]
+    #     set_S!(S, sigma, N)
+    #     P = create_P(rho, x_length)
+    #     mul!(SV, S, V)
+    #     mul!(PV_x, P, V_x)
+    #     @. A = SV - PV_x + R
+    #     # Use the Frobenius norm since it is strictly convex and cheap to compute
+    #     return norm(A)^2
+    # end
+    x0 = zeros(L + N)
+    # opti(x) = optimization_function(x, p)
+    # opti_grad!(G, x) = optimization_function_grad!(G, x, p)
+    fg!(F, G, x) = optimization_function_and_grad!(F, G, x, p)
+    result = optimize(Optim.only_fg!(fg!), x0, LBFGS(), options)
+    # result = optimize(opti, opti_grad!, x0, LBFGS(), options)
+    # result = optimize(opti, x0, LBFGS(), options; autodiff = :forward)
+    display(result)
+    x = minimizer(result)
+    sigma = x[1:L]
+    rho = x[(L + 1):end]
+    S = create_S(sigma, N)
+    P = create_P(rho, x_length)
+    weights = diag(P)
+    Q = S + B/2
+    D = inv(P) * Q
+    return weights, D
+end
+
+@views function optimization_function_and_grad!(F, G, x, p)
+    V, V_x, R, x_length, S_cache, A_cache, SV_cache, PV_x_cache, daij_dsigmak, daij_drhok = p
+    S = get_tmp(S_cache, x)
+    A = get_tmp(A_cache, x)
+    SV = get_tmp(SV_cache, x)
+    PV_x = get_tmp(PV_x_cache, x)
+    (N, _) = size(R)
+    L = div(N * (N - 1), 2)
+    sigma = x[1:L]
+    rho = x[(L + 1):end]
+    set_S!(S, sigma, N)
+    P = create_P(rho, x_length)
+    mul!(SV, S, V)
+    mul!(PV_x, P, V_x)
+    @. A = SV - PV_x + R
+    if !isnothing(G)
+        fill!(daij_dsigmak, zero(eltype(daij_dsigmak)))
+        for k in axes(daij_dsigmak, 3)
+            for j in axes(daij_dsigmak, 2)
+                for i in axes(daij_dsigmak, 1)
+                    l_tilde = k + i - N * (i - 1) + div(i * (i - 1), 2)
+                    # same as above, but needs more type conversions
+                    # l_tilde = Int(k + i - (i - 1) * (N - i/2))
+                    if i + 1 <= l_tilde <= N
+                        daij_dsigmak[i, j, k] += V[l_tilde, j]
+                    else
+                        C = N^2 - 3*N + 2*i - 2*k + 1/4
+                        if C >= 0
+                            D = sqrt(C)
+                            if isinteger(1/2 + D)
+                                l_hat1 = Int(N - 1/2 + D)
+                                l_hat2 = Int(N - 1/2 - D)
```

Comment:
If you want to optimize it further, this is probably the part to look at. Can you get the same results using only integer operations?

If not, maybe you can also reuse some results, e.g., 

```suggestion
                        C = N^2 - 3*N + 2*i - 2*k + 1/4
                        if C >= 0
                            D = sqrt(C)
                            D_plus_one_half = D + 0.5
                            D_plus_one_half_trunc = trunc(D_plus_one_half)
                            if D_plus_one_half == D_plus_one_half_trunc
                                int_D_plus_one_half = trunc(Int, D_plus_one_half_trunc)
                                l_hat1 = N + int_D_plus_one_half - 1
                                l_hat2 = N - int_D_plus_one_half
```

Reply:
Using only integer arithmetic will probably be hard because of the `sqrt`, right? I don't see how we could get rid of that, but let me try your suggestion with reusing some computations.

Reply:
This didn't change so much.